File: src/test/java/boat/info/CloudServiceTest.java
package boat.info;

import java.util.List;

import boat.torrent.Torrent;
import boat.torrent.TorrentFile;
import boat.utilities.PropertiesHelper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class CloudServiceTest {

    private CloudService cloudService;

    @BeforeEach
    public void beforeMethod() {
        cloudService = new CloudService(new CloudFileService());
    }

    @EnabledIfEnvironmentVariable(named = "DISABLE_UPDATE_PROMPT",matches = "true")
    @Test
    void searchForFilesWithYear() {
        // Given
        // When
        List<String> files = cloudService.findExistingFiles("Plan 9 1959");
        // Then
        assertTrue(files.size() > 0);
    }

    @EnabledIfEnvironmentVariable(named = "DISABLE_UPDATE_PROMPT",matches = "true")
    @Test
    void searchForFiles() {
        // Given
        // When
        List<String> files = cloudService.findExistingFiles("Plan 9");
        // Then
        assertTrue(files.size() > 0);
    }

    @Test
    void buildDestinationPathForSingleFileTorrentNoMatch() {
        // Given
        Torrent torrentToBeDownloaded = new Torrent("test");
        // When
        String destinationPath = cloudService.buildDestinationPath("Movie");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/transfer/M/", destinationPath);
    }

    @Test
    void buildDestinationPathForSingleFileTorrent() {
        // Given
        Torrent torrentToBeDownloaded = new Torrent("test");
        // When
        String destinationPath = cloudService.buildDestinationPath("Movie Title Xvid");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Movies/M/", destinationPath);
    }

    @Test
    void buildDestinationPathForSingleFileTorrentWithNumber() {
        // Given
        Torrent torrentToBeDownloaded = new Torrent("test");
        // When
        String destinationPath = cloudService.buildDestinationPath("A 12 Number Title Xvid");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Movies/0-9/", destinationPath);
    }

    @Test
    void buildDestinationPathForSingleFileTorrentExtended() {
        // Given
        Torrent torrentToBeDownloaded = new Torrent("test");
        // When
        String destinationPath = cloudService.buildDestinationPath("Movie Title 2008 2160p US BluRay REMUX HEVC DTS HD MA TrueHD 7 1 Atmos FGT");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Movies/M/", destinationPath);
    }

    @Test
    void buildDestinationPathForSingleFileTorrentWithArticles() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("A Movie Title 2008 2160p US BluRay REMUX HEVC DTS HD MA TrueHD 7 1 Atmos FGT");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Movies/M/", destinationPath);
    }

    @Test
    void buildDestinationPathForMultiFileTorrentTest() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("Test and Test [UNCENSORED] Season 1-3 [1080p] [5.1 MP3] [x265][FINAL]");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/T/Test.And.Test/", destinationPath);
    }

    @Test
    void buildDestinationPathForSeriesTorrentTest() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("Series.S01E02.480p.x264-mSD[tag].mkv");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/S/Series/", destinationPath);
    }

    @Test
    void buildDestinationPathForSeriesLowerCaseTorrentTest() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("series.S01E02.480p.x264-mSD[tag].mkv");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/S/Series/", destinationPath);
    }

    @Test
    void buildDestinationPathWithQuoteTest() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("\"series.S01E02.480p.x264-mSD[tag].mkv");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/S/Series/", destinationPath);
    }

    @Test
    void buildDestinationPathForSeriesTorrentTestSubFolders() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("Series.Name.S01E02.480p.x264-mSD[tag].mkv");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/S/Series.Name/", destinationPath);
    }

    @Test
    void buildDestinationPathWithPDFTest() {
        // Given
        // When
        String destinationPath = cloudService.buildDestinationPath("series.S01E02.480p.x264-mSD[tag].pdf");
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/transfer/S/", destinationPath);
    }

    @Test
    void buildDestinationPathSeriesWithFileListAndNonSeriesTorrentName() {
        // Given
        TorrentFile tf1 = new TorrentFile();
        tf1.name = "Series Name S01E01";
        TorrentFile tf2 = new TorrentFile();
        tf2.name = "Series Name S01E02";
        final Torrent torrent = new Torrent("Test");
        torrent.name = "Could be a movie but is a series";
        // When
        String destinationPath = cloudService.buildDestinationPath(torrent.name, List.of(tf1, tf2));
        // Then
        assertEquals(PropertiesHelper.getProperty("rclonedir") + "/Series-Shows/C/Could.Be.A.Movie.But.Is.A.Series/",
            destinationPath);
    }

}

File: src/test/java/boat/info/CloudServiceTest.kt
package boat.info

import boat.torrent.Torrent
import boat.torrent.TorrentFile
import boat.utilities.HttpHelper
import boat.utilities.PropertiesHelper
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable

internal class CloudServiceTest {
    private var cloudService: CloudService? = null

    @BeforeEach
    fun beforeMethod() {
        cloudService = CloudService(CloudFileService(), TheFilmDataBaseService(HttpHelper()))
    }

    @EnabledIfEnvironmentVariable(named = "DISABLE_UPDATE_PROMPT", matches = "true")
    @Test
    fun searchForFilesWithYear() {
        // Given
        // When
        val files = cloudService!!.findExistingFiles("Plan 9 1959")
        // Then
        Assertions.assertTrue(files.size > 0)
    }

    @EnabledIfEnvironmentVariable(named = "DISABLE_UPDATE_PROMPT", matches = "true")
    @Test
    fun searchForFiles() {
        // Given
        // When
        val files = cloudService!!.findExistingFiles("Plan 9")
        // Then
        Assertions.assertTrue(files.size > 0)
    }

    @Test
    fun buildDestinationPathForSingleFileTorrentNoMatch() {
        // Given
        val torrentToBeDownloaded = Torrent("test")
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Movie", listOf(TorrentFile(name = "movie.pdf"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/transfer/M/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSingleFileTorrent() {
        // Given
        val torrentToBeDownloaded = Torrent("test")
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Movie Title Xvid", listOf(TorrentFile(name = "xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Movies/M/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSingleFileTorrentWithNumber() {
        // Given
        val torrentToBeDownloaded = Torrent("test")
        // When
        val destinationPath = cloudService!!.buildDestinationPath("A 12 Number Title Xvid", listOf(TorrentFile(name = "xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Movies/0-9/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSingleFileTorrentExtended() {
        // Given
        val torrentToBeDownloaded = Torrent("test")
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Movie Title 2008 2160p US BluRay REMUX HEVC DTS HD MA TrueHD 7 1 Atmos FGT", listOf(TorrentFile(name = "xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Movies/M/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSingleFileTorrentWithArticles() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("A Movie Title 2008 2160p US BluRay REMUX HEVC DTS HD MA TrueHD 7 1 Atmos FGT", listOf(TorrentFile(name = "xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Movies/M/", destinationPath)
    }

    @Test
    fun buildDestinationPathForMultiFileTorrentTest() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Test and Test [UNCENSORED] Season 1-3 [1080p] [5.1 MP3] [x265][FINAL]", listOf(TorrentFile(name = "Season 1-3 xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/T/Test.And.Test/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSeriesTorrentTest() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Series.S01E02.480p.x264-mSD[tag].mkv", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSeriesLowerCaseTorrentTest() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("series.S01E02.480p.x264-mSD[tag].mkv", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series/", destinationPath)
    }

    @Test
    fun buildDestinationPathWithQuoteTest() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("\"series.S01E02.480p.x264-mSD[tag].mkv", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series/", destinationPath)
    }

    @Test
    fun buildDestinationPathForSeriesTorrentTestSubFolders() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("Series.Name.S01E02.480p.x264-mSD[tag].mkv", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series.Name/", destinationPath)
    }

    @Test
    fun buildDestinationPathWithPDFTest() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("series.S01E02.480p.x264-mSD[tag].pdf", listOf(TorrentFile(name = "S01E02xvid.pdf"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/transfer/S/", destinationPath)
    }

    @Test
    fun buildDestinationPathSeriesWithFileListAndNonSeriesTorrentName() {
        // Given
        val tf1 = TorrentFile()
        tf1.name = "Series Name S01E01.mkv"
        val tf2 = TorrentFile()
        tf2.name = "Series Name S01E02.mkv"
        val torrent = Torrent("Test")
        torrent.name = "Could be a movie but is a series"
        // When
        val destinationPath = cloudService!!.buildDestinationPath(torrent.name, listOf(tf1, tf2)).second
        // Then
        Assertions.assertEquals(
            PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/C/Could.Be.A.Movie.But.Is.A.Series/",
            destinationPath
        )
    }

    @Test
    fun buildDestinationPathWithSeriesAndExtraTextAfterSeriesName() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("series.S01E02.480p.x264-mSD[tag]", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series/", destinationPath)
    }

    @Test
    fun buildDestinationPathWithSeriesAndExtraTextAfterSeriesNameExtended() {
        // Given
        // When
        val destinationPath = cloudService!!.buildDestinationPath("series.S01E02.Test.480p.x264-mSD[tag]", listOf(TorrentFile(name = "S01E02xvid.xvid"))).second
        // Then
        Assertions.assertEquals(PropertiesHelper.getProperty("RCLONEDIR") + "/Series-Shows/S/Series/", destinationPath)
    }
}